<a href="https://colab.research.google.com/github/carlosdgerez/machine_learning/blob/main/secondBoost.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd

housing = pd.read_csv('https://raw.githubusercontent.com/byui-cse/cse450-course/master/data/housing.csv')

In [3]:
#Changing date to be in the same format as the rest of the years features
import datetime
housing['date'] = pd.DatetimeIndex(housing['date']).year



In [4]:
# selecting clean dat including lat and long
clean_data = housing.drop(['id'], axis = 1)

In [ ]:
clean_data.head()

date  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  view  \
0  2014         4       3.25         3760      4675     2.0           0     0   
1  2015         3       1.75         1460      7800     1.0           0     0   
2  2014         4       1.00         1340      5800     1.5           0     2   
3  2014         3       1.75         1440      8050     1.0           0     0   
4  2014         2       1.50         1780      4750     1.0           0     0   

   condition  grade  sqft_above  sqft_basement  yr_built  yr_renovated  \
0          3      8        2740           1020      2007             0   
1          2      7        1040            420      1979             0   
2          3      7        1340              0      1914             0   
3          3      8        1440              0      1985             0   
4          4      7        1080            700      1947             0   

   zipcode      lat     long  sqft_living15  sqft_lot15     price  
0    98038  47.3862 -122.048           3280        4033  429900.0  
1    98023  47.3035 -122.382           1310        7865  233000.0  
2    98116  47.5658 -122.389           1900        5800  455000.0  
3    98023  47.3187 -122.390           1790        7488  258950.0  
4    98117  47.6859 -122.395           1690        5962  555000.0

In [5]:

# change date name to something more descriptive
clean_data.rename(columns={'date': 'sale_date'}, inplace=True)

In [6]:
# get a training set: a X as with the features to relate and a y with the  target 
X = clean_data.iloc[:,0:-1]
y = clean_data.price


In [ ]:
X

sale_date  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0           2014         4       3.25         3760      4675     2.0   
1           2015         3       1.75         1460      7800     1.0   
2           2014         4       1.00         1340      5800     1.5   
3           2014         3       1.75         1440      8050     1.0   
4           2014         2       1.50         1780      4750     1.0   
...          ...       ...        ...          ...       ...     ...   
19995       2014         3       1.50         1000      6914     1.0   
19996       2014         3       2.50         3087      5002     2.0   
19997       2014         3       2.50         2120      4780     2.0   
19998       2014         1       0.75          380     15000     1.0   
19999       2015         4       2.50         3130      5999     2.0   

       waterfront  view  condition  grade  sqft_above  sqft_basement  \
0               0     0          3      8        2740           1020   
1               0     0          2      7        1040            420   
2               0     2          3      7        1340              0   
3               0     0          3      8        1440              0   
4               0     0          4      7        1080            700   
...           ...   ...        ...    ...         ...            ...   
19995           0     0          3      7        1000              0   
19996           0     0          3      8        3087              0   
19997           0     0          3      7        2120              0   
19998           0     0          3      5         380              0   
19999           0     0          3      7        3130              0   

       yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0          2007             0    98038  47.3862 -122.048           3280   
1          1979             0    98023  47.3035 -122.382           1310   
2          1914             0    98116  47.5658 -122.389           1900   
3          1985             0    98023  47.3187 -122.390           1790   
4          1947             0    98117  47.6859 -122.395           1690   
...         ...           ...      ...      ...      ...            ...   
19995      1947             0    98125  47.7144 -122.319           1000   
19996      2014             0    98023  47.2974 -122.349           2927   
19997      2004             0    98053  47.6810 -122.032           1690   
19998      1963             0    98168  47.4810 -122.323           1170   
19999      2006             0    98042  47.3837 -122.099           3020   

       sqft_lot15  
0            4033  
1            7865  
2            5800  
3            7488  
4            5962  
...           ...  
19995        6947  
19996        5183  
19997        2650  
19998       15000  
19999        5997  

[20000 rows x 19 columns]

In [7]:
# Start spliting the data and import libraries

#import train split
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=.30, random_state=1)


In [8]:
#  xgboost regresor
import xgboost
from sklearn import metrics as sm

# createn instance and fit the model

clf2 = xgboost.XGBRegressor(random_state=5, max_depth= 4, n_estimators= 900, learning_rate= 0.1, verbosity = 0, n_jobs = -1).fit(X_train, y_train)
y_pred = clf2.predict(X_test)

# This is how to generate a regresion report ( for regresion )
print("Mean absolute error =", round(
    sm.mean_absolute_error(y_test, y_pred), 2))
print("Mean squared error =", round(
    sm.mean_squared_error(y_test, y_pred), 2))
print("Median absolute error =", round(
    sm.median_absolute_error(y_test, y_pred), 2))
print("Explain variance score =", round(
    sm.explained_variance_score(y_test, y_pred), 2))
print("R2 score =", round( sm.r2_score(y_test, y_pred),4))
print( "ROOT Mean squared error = ", 
    round(sm.mean_squared_error(y_test, y_pred, squared=False 
)))


Mean absolute error = 63456.75
Mean squared error = 12521655714.93
Median absolute error = 37896.84
Explain variance score = 0.91
R2 score = 0.9061
ROOT Mean squared error =  111900


## Searching multiple parameters simultaneously

In [ ]:
from sklearn.model_selection import GridSearchCV

In [ ]:
# create list with the value for the parameters check

estimators = [100, 600, 900]
depth = range(2,6)
learning = [0.1,0.05, 0.01]

In [ ]:
# Create the parameters grid dictionary
param_grid = dict(n_estimators=estimators, max_depth= depth, learning_rate= learning)


In [ ]:
# instatiate the grid
grid = GridSearchCV(clf2, param_grid, cv=10, scoring='r2')

In [ ]:
# fit the grid with data remember just X and y THIS CODE TAKES AT LEAST 15 to 30 MINUTES TO RUN
grid.fit(X,y)

GridSearchCV(cv=10,
             estimator=XGBRegressor(max_depth=4, n_estimators=600,
                                    random_state=5, verbosity=0),
             param_grid={'learning_rate': [0.1, 0.05, 0.01],
                         'max_depth': range(2, 6),
                         'n_estimators': [100, 600, 900]},
             scoring='r2')

In [ ]:
print(grid.best_score_)
print (grid.best_params_)

0.9022325723439859
{'learning_rate': 0.1, 'max_depth': 4, 'n_estimators': 900}


In [ ]:
from sklearn.model_selection import RandomizedSearchCV

In [ ]:
param_dist = dict(n_estimators=estimators, max_depth = depth, learning_rate= learning)
rand = RandomizedSearchCV (clf2, param_dist, cv=10, scoring='r2', n_iter=10, random_state=5, n_jobs=-1)
rand.fit(X,y)

RandomizedSearchCV(cv=10,
                   estimator=XGBRegressor(max_depth=4, n_estimators=600,
                                          random_state=5, verbosity=0),
                   param_distributions={'learning_rate': [0.1, 0.05, 0.01],
                                        'max_depth': range(3, 6),
                                        'n_estimators': [100, 600, 900]},
                   random_state=5, scoring='r2')

In [ ]:
print(rand.best_score_)

0.8946050187521459


In [ ]:
print(rand.best_score_)
print(rand.best_params_)

0.8946050187521459
{'n_estimators': 900, 'max_depth': 3, 'learning_rate': 0.1}


# Adding new data to the set to improve performance

In [9]:
# download the income data

zip_code = pd.read_csv('https://raw.githubusercontent.com/carlosdgerez/machine_learning/main/mean_income_washington_by_zipcode.csv')
zip_code.head()

zip_code  mean_income
0     98001       102586
1     98002        70945
2     98003        76753
3     98004       210129
4     98005       186020

In [10]:
# rename the columns as the one in clean data to use merge by this column
zip_code.rename(columns={'zip_code': 'zipcode'}, inplace = True)

In [11]:
# Use merge to get the data in the dataframe
new_clean_data = pd.merge(clean_data, zip_code, how= 'left', on= 'zipcode')

In [12]:
# reorder the data to have the target price at the end
column_to_move = new_clean_data.pop("mean_income")

# insert column with insert(location, column_name, column_value)

new_clean_data.insert(19, "mean_income", column_to_move)


# Run the model with the best parameters to see if improve the performance
this is the best model until now

In [13]:
# get a training set: a X as with the features to relate and a y with the  target 
X = new_clean_data.iloc[:,0:-1]
y = new_clean_data.price


In [14]:
# Start spliting the data and import libraries



X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=.30, random_state=1)


In [15]:
#  xgboost regresor


# createn instance and fit the model (runing time 12 seconds aprox)

clf2 = xgboost.XGBRegressor(random_state=5, max_depth= 4, n_estimators= 900, learning_rate= 0.1, verbosity = 0, n_jobs=-1).fit(X_train, y_train)
y_pred = clf2.predict(X_test)

# This is how to generate a regresion report ( for regresion )
print("Mean absolute error in dollars =", round(
    sm.mean_absolute_error(y_test, y_pred), 2))
print("Mean squared error =", round(
    sm.mean_squared_error(y_test, y_pred), 2))
print("Median absolute error =", round(
    sm.median_absolute_error(y_test, y_pred), 2))
print("Explain variance score =", round(
    sm.explained_variance_score(y_test, y_pred), 2))
print("R2 score =", round( sm.r2_score(y_test, y_pred),4))
print( "ROOT Mean squared error in dollars = ", 
    round(sm.mean_squared_error(y_test, y_pred, squared=False 
)))



Mean absolute error in dollars = 62444.12
Mean squared error = 12296249175.04
Median absolute error = 37460.83
Explain variance score = 0.91
R2 score = 0.9078
ROOT Mean squared error in dollars =  110888


In [16]:


#  EVALUATION OF THE MODEL 

#Predict the response for test dataset


# Features importances in a model this block can be run with diferent models to compare


feat_imports = (pd.DataFrame({"feature names": X_train.columns,
                              "importances": clf2.feature_importances_})
                .sort_values("importances", ascending=False))

feat_imports.reset_index(drop= True, inplace= True)



In [ ]:
new_features = feat_imports['feature names'][0:10]

In [ ]:
new_features

0            grade
1       waterfront
2      sqft_living
3      mean_income
4              lat
5             view
6             long
7         yr_built
8    sqft_living15
9          zipcode
Name: feature names, dtype: object

# Run model with less features

In [ ]:
# get a training set: a X as with the features to relate and a y with the  target 
X = new_clean_data[new_features]
y = new_clean_data.price


In [ ]:
# Start spliting the data and import libraries



X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=.30, random_state=1)


In [ ]:
#  xgboost regresor


# createn instance and fit the model

clf3 = xgboost.XGBRegressor(random_state=5, max_depth= 4, n_estimators= 900, learning_rate= 0.1, verbosity = 0, n_jobs=-1).fit(X_train, y_train)
y_pred = clf3.predict(X_test)

# This is how to generate a regresion report ( for regresion )
print("Mean absolute error in dollars =", round(
    sm.mean_absolute_error(y_test, y_pred), 2))
print("Mean squared error =", round(
    sm.mean_squared_error(y_test, y_pred), 2))
print("Median absolute error =", round(
    sm.median_absolute_error(y_test, y_pred), 2))
print("Explain variance score =", round(
    sm.explained_variance_score(y_test, y_pred), 2))
print("R2 score =", round( sm.r2_score(y_test, y_pred),4))
print( "ROOT Mean squared error in dollars = ", 
    round(sm.mean_squared_error(y_test, y_pred, squared=False 
)))



Mean absolute error in dollars = 68332.42
Mean squared error = 15009879103.6
Median absolute error = 41699.44
Explain variance score = 0.89
R2 score = 0.8875
ROOT Mean squared error in dollars =  122515


Reducing the name of columns by importances yield worst results

Graphing importances with altair. This features importances are from the best scored model clf2

In [17]:
import altair as alt
alt.Chart(feat_imports.head(10)).mark_bar().encode(
    alt.X("importances", axis=alt.Axis(format='%')),
    alt.Y('feature names',sort='-x')
)

alt.Chart(...)

In [ ]:
new_clean_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20000 entries, 0 to 19999
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   sale_date      20000 non-null  int64  
 1   bedrooms       20000 non-null  int64  
 2   bathrooms      20000 non-null  float64
 3   sqft_living    20000 non-null  int64  
 4   sqft_lot       20000 non-null  int64  
 5   floors         20000 non-null  float64
 6   waterfront     20000 non-null  int64  
 7   view           20000 non-null  int64  
 8   condition      20000 non-null  int64  
 9   grade          20000 non-null  int64  
 10  sqft_above     20000 non-null  int64  
 11  sqft_basement  20000 non-null  int64  
 12  yr_built       20000 non-null  int64  
 13  yr_renovated   20000 non-null  int64  
 14  zipcode        20000 non-null  int64  
 15  lat            20000 non-null  float64
 16  long           20000 non-null  float64
 17  sqft_living15  20000 non-null  int64  
 18  sqft_l

Add new data with price index variations for the zip codes we are working with

In [ ]:
# download the income data

price_index = pd.read_csv('https://raw.githubusercontent.com/carlosdgerez/machine_learning/main/price_index.csv')
price_index.head()

zipcode  Annual Change (%)      HPI  HPI with 1990 base  HPI with 2000 base
0    98001              16.18  1175.72              378.94              255.89
1    98002              15.31  1061.13              424.05              263.15
2    98003              17.35  1241.23              393.20              274.01
3    98004             -10.87  1876.50              473.89              273.74
4    98005              -0.55  1646.14              467.74              288.62

In [ ]:
# Use merge to get the data in the dataframe
new_clean_data2 = pd.merge(new_clean_data, price_index, how= 'left', on= 'zipcode')

In [ ]:
new_clean_data2.head()


sale_date  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  \
0       2014         4       3.25         3760      4675     2.0           0   
1       2015         3       1.75         1460      7800     1.0           0   
2       2014         4       1.00         1340      5800     1.5           0   
3       2014         3       1.75         1440      8050     1.0           0   
4       2014         2       1.50         1780      4750     1.0           0   

   view  condition  grade  ...      lat     long  sqft_living15  sqft_lot15  \
0     0          3      8  ...  47.3862 -122.048           3280        4033   
1     0          2      7  ...  47.3035 -122.382           1310        7865   
2     2          3      7  ...  47.5658 -122.389           1900        5800   
3     0          3      8  ...  47.3187 -122.390           1790        7488   
4     0          4      7  ...  47.6859 -122.395           1690        5962   

   mean_income     price  Annual Change (%)      HPI  HPI with 1990 base  \
0       134178  429900.0              14.75  1140.60              391.43   
1       100910  233000.0              16.22  1271.45              372.36   
2       151980  455000.0               5.44  2213.57              489.15   
3       100910  258950.0              16.22  1271.45              372.36   
4       166123  555000.0               8.18  2345.21              545.24   

   HPI with 2000 base  
0              254.48  
1              264.66  
2              271.53  
3              264.66  
4              309.96  

[5 rows x 25 columns]

In [ ]:
# reorder the data to have the target price at the end
column_to_move = new_clean_data2.pop("price")

# insert column with insert(location, column_name, column_value)

new_clean_data2.insert(24, "price", column_to_move)


In [ ]:
new_clean_data2.head()

sale_date  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  \
0       2014         4       3.25         3760      4675     2.0           0   
1       2015         3       1.75         1460      7800     1.0           0   
2       2014         4       1.00         1340      5800     1.5           0   
3       2014         3       1.75         1440      8050     1.0           0   
4       2014         2       1.50         1780      4750     1.0           0   

   view  condition  grade  ...      lat     long  sqft_living15  sqft_lot15  \
0     0          3      8  ...  47.3862 -122.048           3280        4033   
1     0          2      7  ...  47.3035 -122.382           1310        7865   
2     2          3      7  ...  47.5658 -122.389           1900        5800   
3     0          3      8  ...  47.3187 -122.390           1790        7488   
4     0          4      7  ...  47.6859 -122.395           1690        5962   

   mean_income  Annual Change (%)      HPI  HPI with 1990 base  \
0       134178              14.75  1140.60              391.43   
1       100910              16.22  1271.45              372.36   
2       151980               5.44  2213.57              489.15   
3       100910              16.22  1271.45              372.36   
4       166123               8.18  2345.21              545.24   

   HPI with 2000 base     price  
0              254.48  429900.0  
1              264.66  233000.0  
2              271.53  455000.0  
3              264.66  258950.0  
4              309.96  555000.0  

[5 rows x 25 columns]

At this point I have a new dataset call new_clean_data2 and can start to test in the model I have clf2

In [ ]:
# get a training set: a X as with the features to relate and a y with the  target 
X = new_clean_data2.iloc[:,0:-1]
y = new_clean_data2.price


In [ ]:
# Start spliting the data and import libraries



X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=.30, random_state=1)


In [ ]:
#  xgboost regresor


# createn instance and fit the model (runing time 12 seconds aprox)

clf2 = xgboost.XGBRegressor(random_state=5, max_depth= 4, n_estimators= 900, learning_rate= 0.1, verbosity = 0, n_jobs=-1).fit(X_train, y_train)
y_pred = clf2.predict(X_test)

# This is how to generate a regresion report ( for regresion )
print("Mean absolute error in dollars =", round(
    sm.mean_absolute_error(y_test, y_pred), 2))
print("Mean squared error =", round(
    sm.mean_squared_error(y_test, y_pred), 2))
print("Median absolute error =", round(
    sm.median_absolute_error(y_test, y_pred), 2))
print("Explain variance score =", round(
    sm.explained_variance_score(y_test, y_pred), 2))
print("R2 score =", round( sm.r2_score(y_test, y_pred),4))
print( "ROOT Mean squared error in dollars = ", 
    round(sm.mean_squared_error(y_test, y_pred, squared=False 
)))



Mean absolute error in dollars = 61570.64
Mean squared error = 12337913661.56
Median absolute error = 36403.17
Explain variance score = 0.91
R2 score = 0.9075
ROOT Mean squared error in dollars =  111076


Not an improvement here, the r2 its 0.9075, and without this data was 0.9078.

# New features to try

In [19]:

zipcode = pd.read_csv('https://raw.githubusercontent.com/Kate-Strydom/cse450/main/notebooks/zipcode_washington.csv')

zipcode.head()

zipcode  population  median_household_income
0    98001       31742                    72540
1    98002       34240                    44930
2    98003       45632                    45563
3    98004       31254                   100344
4    98005       17880                    85639

In [25]:
# Use merge to get the data in the dataframe
new_clean_data3 = pd.merge(new_clean_data, zipcode, how= 'left', on= 'zipcode')

In [28]:
new_clean_data3.head()

sale_date  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  \
0       2014         4       3.25         3760      4675     2.0           0   
1       2015         3       1.75         1460      7800     1.0           0   
2       2014         4       1.00         1340      5800     1.5           0   
3       2014         3       1.75         1440      8050     1.0           0   
4       2014         2       1.50         1780      4750     1.0           0   

   view  condition  grade  ...  yr_renovated  zipcode      lat     long  \
0     0          3      8  ...             0    98038  47.3862 -122.048   
1     0          2      7  ...             0    98023  47.3035 -122.382   
2     2          3      7  ...             0    98116  47.5658 -122.389   
3     0          3      8  ...             0    98023  47.3187 -122.390   
4     0          4      7  ...             0    98117  47.6859 -122.395   

   sqft_living15  sqft_lot15  mean_income  population  \
0           3280        4033       134178       34076   
1           1310        7865       100910       49402   
2           1900        5800       151980       23780   
3           1790        7488       100910       49402   
4           1690        5962       166123       31778   

   median_household_income     price  
0                    93573  429900.0  
1                    67881  233000.0  
2                    73667  455000.0  
3                    67881  258950.0  
4                    91199  555000.0  

[5 rows x 23 columns]

In [27]:
# reorder the data to have the target price at the end
column_to_move = new_clean_data3.pop("price")

# insert column with insert(location, column_name, column_value)

new_clean_data3.insert(22, "price", column_to_move)


In [29]:
# get a training set: a X as with the features to relate and a y with the  target 
X = new_clean_data3.iloc[:,0:-1]
y = new_clean_data3.price


In [30]:
# Start spliting the data and import libraries



X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=.30, random_state=1)


In [35]:
#  xgboost regresor


# createn instance and fit the model (runing time 12 seconds aprox)

clf4 = xgboost.XGBRegressor(random_state=5, max_depth= 4, n_estimators= 900, learning_rate= 0.1, verbosity = 0, n_jobs=-1).fit(X_train, y_train)
y_pred = clf4.predict(X_test)

# This is how to generate a regresion report ( for regresion )
print("Mean absolute error in dollars =", round(
    sm.mean_absolute_error(y_test, y_pred), 2))
print("Mean squared error =", round(
    sm.mean_squared_error(y_test, y_pred), 2))
print("Median absolute error =", round(
    sm.median_absolute_error(y_test, y_pred), 2))
print("Explain variance score =", round(
    sm.explained_variance_score(y_test, y_pred), 2))
print("R2 score =", round( sm.r2_score(y_test, y_pred),4))
print( "ROOT Mean squared error in dollars = ", 
    round(sm.mean_squared_error(y_test, y_pred, squared=False 
)))



Mean absolute error in dollars = 61911.21
Mean squared error = 11884782245.04
Median absolute error = 36993.66
Explain variance score = 0.91
R2 score = 0.9109
ROOT Mean squared error in dollars =  109017


In [36]:


#  EVALUATION OF THE MODEL 

#Predict the response for test dataset


# Features importances in a model this block can be run with diferent models to compare


feat_imports = (pd.DataFrame({"feature names": X_train.columns,
                              "importances": clf4.feature_importances_})
                .sort_values("importances", ascending=False))

feat_imports.reset_index(drop= True, inplace= True)



In [57]:
import altair as alt

feat_imports['percents'] = round((feat_imports.importances * 100).astype(float),2)
alt.Chart(feat_imports.head(20)).mark_bar().encode(
    alt.X("importances", axis=alt.Axis(format='%')),
    alt.Y('feature names',sort='-x'),
    tooltip=('percents'))

alt.Chart(...)

In [54]:
feat_imports

feature names  importances  percents
0                     grade     0.328326        32
1                waterfront     0.208753        20
2               sqft_living     0.130076        13
3               mean_income     0.074025         7
4                       lat     0.051398         5
5                      view     0.047247         4
6                      long     0.041043         4
7                  yr_built     0.019437         1
8                   zipcode     0.015727         1
9                 bathrooms     0.013580         1
10            sqft_living15     0.013185         1
11               sqft_above     0.011407         1
12               population     0.007556         0
13                 sqft_lot     0.005428         0
14                condition     0.005397         0
15             yr_renovated     0.004709         0
16                   floors     0.004591         0
17            sqft_basement     0.004145         0
18               sqft_lot15     0.004090         0
19  median_household_income     0.003796         0
20                sale_date     0.003602         0
21                 bedrooms     0.002480         0